src/lib/permutations/keccak_fips/keccak_fips.cpp

```diff
@@ -0,0 +1,179 @@
+
+   for(size_t i = 0; i != 24; i += 2) {
+      Keccak_FIPS_round(T, A, RC[i + 0]);
+      Keccak_FIPS_round(A, T, RC[i + 1]);
+   }
+}
+
+void Keccak_FIPS::permute() {
+   Keccak_FIPS::permute(m_S.data());
+}
+
+//static
+
+uint32_t Keccak_FIPS::absorb(size_t bitrate,
+                             secure_vector<uint64_t>& S,
+                             size_t S_pos,
+                             std::span<const uint8_t> input_span) {
+   const uint8_t* input = input_span.data();
+   size_t length = input_span.size();
+
+   while(length > 0) {
+      size_t to_take = std::min(length, bitrate / 8 - S_pos);
+
+      length -= to_take;
+
+      while(to_take && S_pos % 8) {
+         S[S_pos / 8] ^= static_cast<uint64_t>(input[0]) << (8 * (S_pos % 8));
+
+         ++S_pos;
+         ++input;
+         --to_take;
+      }
+
+      while(to_take && to_take % 8 == 0) {
+         S[S_pos / 8] ^= load_le<uint64_t>(input, 0);
+         S_pos += 8;
+         input += 8;
+         to_take -= 8;
+      }
+
+      while(to_take) {
+         S[S_pos / 8] ^= static_cast<uint64_t>(input[0]) << (8 * (S_pos % 8));
+
+         ++S_pos;
+         ++input;
+         --to_take;
+      }
+
+      if(S_pos == bitrate / 8) {
+         Keccak_FIPS::permute(S.data());
+         S_pos = 0;
+      }
+   }
+
+   return static_cast<uint32_t>(S_pos);
+}
+
+//static
+
+void Keccak_FIPS::finish(
+   size_t bitrate, secure_vector<uint64_t>& S, size_t S_pos, uint64_t custom_padd, uint8_t custom_padd_bit_len) {
+   BOTAN_ARG_CHECK(bitrate % 64 == 0, "Keccak-FIPS bitrate must be multiple of 64");
+   // append the first bit of the final padding after the custom padding
+   uint8_t init_pad = static_cast<uint8_t>(custom_padd | uint64_t(1) << custom_padd_bit_len);
+   S[S_pos / 8] ^= static_cast<uint64_t>(init_pad) << (8 * (S_pos % 8));
+   // final bit of the padding of the last block
+   S[(bitrate / 64) - 1] ^= static_cast<uint64_t>(0x80) << 56;
+   Keccak_FIPS::permute(S.data());
+}
+
+//static
+
+void Keccak_FIPS::expand(size_t bitrate, secure_vector<uint64_t>& S, std::span<uint8_t> output_span) {
+   uint8_t* output = output_span.data();
+   size_t output_length = output_span.size();
+   BOTAN_ARG_CHECK(bitrate % 64 == 0, "Keccak-FIPS bitrate must be multiple of 64");
+
+   const size_t byterate = bitrate / 8;
+
+   while(output_length > 0) {
+      const size_t copying = std::min(byterate, output_length);
+
+      copy_out_vec_le(output, copying, S);
+
+      output += copying;
+      output_length -= copying;
+
+      if(output_length > 0) {
+         Keccak_FIPS::permute(S.data());
+      }
+   }
+}
+
+void Keccak_FIPS::expand(std::span<uint8_t> output_span) {
+   expand(m_bitrate, m_S, output_span);
+}
+
+Keccak_FIPS::Keccak_FIPS(size_t output_bits, size_t capacity, uint64_t custom_padd, uint8_t custom_padd_bit_len) :
+      m_output_bits(output_bits),
+      m_capacity(static_cast<uint32_t>(capacity)),
+      m_bitrate(static_cast<uint32_t>(1600 - capacity)),
```

Comment:
`m_capacity` and `m_bitrate` should just be `size_t`.

The general code guideline - which apparently is not mentioned anywhere in the docs - is that "integers of no specific size" are `size_t`. A variable should be `uintX_t` only because if that variable was any other size/type it would be incorrect.

Reply:
Addressed in #3675.